core/src/main/java/feign/Contract.java

```diff
@@ -211,99 +207,224 @@ protected void nameParam(MethodMetadata data, String name, int i) {
     }
   }
 
-  class Default extends BaseContract {
+  /**
+   * {@link Contract} base implementation that works by declaring witch annotations should be
+   * processed and how each annotation modifies {@link MethodMetadata}
+   */
+  abstract class DeclarativeContract extends BaseContract {
 
-    static final Pattern REQUEST_LINE_PATTERN = Pattern.compile("^([A-Z]+)[ ]*(.*)$");
+    Map<Class<Annotation>, ClassAnnotationProcessor<Annotation>> classAnnotationProcessors =
+        new HashMap<>();
+    private Map<Class<Annotation>, MethodAnnotationProcessor<Annotation>> methodAnnotationProcessors =
+        new HashMap<>();
+    Map<Class<Annotation>, ParameterAnnotationProcessor<Annotation>> parameterAnnotationProcessors =
```

Comment:
this maps are awful, and kind useless, cause I'm force to loop the entries to figure out if an annotation is instance of class.  Need a better data structure to hold this data, but, I ran out of ideas.

Also, probably a single data structure would be enough, no need for one per processor.

Reply:
I don’t think we need explicit maps. The annotations declare where they can be used and if we scan the annotations on either a class method or parameter, we will only see valid ones.

Reply:
my concern was the same annotation being present on class and method, but, demanding different logic to process...

for instance jaxrs `@Path`... 
when you annotate class you will set that as uri
but, when you annotate method, you need to append.... get it could be the same code, but there can be other cases where it may not be

Reply:
I did something similar in Feignx. I delegated that responsibility to MethodMetadata.  I added a guard in the annotation processor to check the state of MethodMetadata and then make the appropriate modification. I’ll add more context once I can get back to my pc.

Reply:
I guess I could go ahead, change all contracts to be declarative.

If all can work with a single code path I could squash this into a single map.